### ISSUE TITLE
libnetdata is not usable from c++ (cpp)

### ISSUE THREAD

**Issue by KickerTom** [netdata/netdata#10178]
##### Bug report summary
Attempting to write a new plugin for netdata in C++, I found that the headers are C only, unusable for C++ application. I understand netdata is (partially) C project, but interface libraries, and especially headers, could be C++ clean to make it more usable.
One of the most prominent problems is usage of _this_ as a struct member, while _this_ is reserved keyword in cpp.

##### Component Name
libnetdata

##### Steps To Reproduce
1. Create simple test.cpp file including libnetdata.h, e.g.:
```
#include "config.h"
#include "libnetdata/libnetdata.h"

int main(int argc, char **argv)
{
    return 0;
}
```
2. Try compiling the code with e.g.
`g++ -DTARGET_OS=OS_LINUX test.cpp`

3. You will receive many errors:
```
In file included from libnetdata/libnetdata.h:304,
                 from test.cpp:2:
libnetdata/avl/avl.h:43:14: error: declaration of ‘avl_tree avl_tree_lock::avl_tree’ [-fpermissive]
     avl_tree avl_tree;
              ^~~~~~~~
libnetdata/avl/avl.h:40:3: error: changes meaning of ‘avl_tree’ from ‘typedef struct avl_tree avl_tree’ [-fpermissive]
 } avl_tree;
   ^~~~~~~~
In file included from libnetdata/libnetdata.h:313,
                 from test.cpp:2:
libnetdata/config/appconfig.h:114:9: error: declaration of ‘avl config_option::avl’ [-fpermissive]
     avl avl;                // the index entry of this entry - this has to be first!
         ^~~
In file included from libnetdata/libnetdata.h:304,
                 from test.cpp:2:
libnetdata/avl/avl.h:34:3: error: changes meaning of ‘avl’ from ‘typedef struct avl avl’ [-fpermissive]
 } avl;
   ^~~
In file included from libnetdata/libnetdata.h:313,
                 from test.cpp:2:
libnetdata/config/appconfig.h:127:9: error: declaration of ‘avl section::avl’ [-fpermissive]
     avl avl;                // the index entry of this section - this has to be first!
         ^~~
In file included from libnetdata/libnetdata.h:304,
                 from test.cpp:2:
libnetdata/avl/avl.h:34:3: error: changes meaning of ‘avl’ from ‘typedef struct avl avl’ [-fpermissive]
 } avl;
   ^~~
In file included from libnetdata/libnetdata.h:316,
                 from test.cpp:2:
libnetdata/dictionary/dictionary.h:16:9: error: declaration of ‘avl name_value::avl’ [-fpermissive]
     avl avl;                // the index - this has to be first!
         ^~~
In file included from libnetdata/libnetdata.h:304,
                 from test.cpp:2:
libnetdata/avl/avl.h:34:3: error: changes meaning of ‘avl’ from ‘typedef struct avl avl’ [-fpermissive]
 } avl;
   ^~~
In file included from libnetdata/libnetdata.h:320,
                 from test.cpp:2:
libnetdata/eval/eval.h:31:24: error: expected unqualified-id before ‘this’
     calculated_number *this;
                        ^~~~
```

##### Expected behavior
The C code from libnetdata should be usable from C++.

I am willing to do some cleanup and submit patches, but I want to know if such (potentially disruptive) change would be acceptable.

**Comment by thiagoftsm** [comment]
Hello @KickerTom ,

I am a long time without to write `C++` code, but I agree with everything you wrote, so I would like your help to refresh my memory, would be enough to add:

```
# ifdef __cplusplus
extern "C" {
# endif

....

# ifdef __cplusplus
}
# endif
```

inside our headers?

Best regards!

**Comment by KickerTom** [comment]
Unfortunately, it's not that easy. Extern "C" clause helps to get function signatures without decorations, so they can be linked with C++ code. That's needed to make the headers usable, but that can also be done from outside, in the worst case.
e.g.:
```
extern "C" {
#include <libnetdata.h>
 }
```
But that can be the last step. What needs to be done first is to make the code C++ clean. E.g. the mentioned **this** in _libnetdata/eval/eval.h_ is a reserved word in C++, co it can't be used as a struct member. So we need to rename it (which may break code using these structures, if they are living outside of netdata repository, so we can't fix it along).

I will try preparing a MR to show the amount of changes needed. I admit I haven't investigated the issue in depth yet, I stopped as soon as I found the **this** member ;-).

**Comment by KickerTom** [comment]
After a quick first check, I found two distinct issues. First is the mentioned **this** struct member, which needs to be renamed. Second is multiple cases when type has the same name as its instance (which is not allowed in C++). This is the case for avl and avl_tree structures. I will open MR to show this in detail, then we can discuss possible solution.

**Comment by thiagoftsm** [comment]
All right @KickerTom .

I already assigned myself and other developers that will be reviewers of your work, when you finish, it is only necessary to remove the `draft` from the PR.

I will also do another request, please add `Fixes #10178` for your PR, this will close this issue when it is merged.

Thank you to bring this fix for us!

Best regards!

**Comment by thiagoftsm** [comment]
I am closing this issue after to merge #10185.
